modules/300-prometheus/openapi/config-values.yaml

```diff
@@ -112,6 +112,19 @@ properties:
       The class of the Ingress controller used for Grafana/Prometheus.
 
       An optional parameter. By default, the `modules.ingressClass` global value is used.
+  externalLabels:
+    type: object
+    x-examples:
+      - prometheus: dev
+        prometheus_replica: ${HOSTNAME}
+        env: $MY_ENV
```

Comment:
This does not seem right.

Users cannot specify any env variables, only a limited subset of envs of the reloader container.

The example os envs from my cluster:
```
KUBERNETES_PORT=tcp://10.222.0.1:443
KUBERNETES_SERVICE_PORT=443
GRAFANA_DEX_AUTHENTICATOR_PORT_443_TCP_ADDR=10.222.133.254
CONTROL_PLANE_PROXY_SERVICE_HOST=10.222.191.230
CONTROL_PLANE_PROXY_PORT_8080_TCP_ADDR=10.222.191.230
HOSTNAME=prometheus-main-0
GRAFANA_DEX_AUTHENTICATOR_PORT_443_TCP_PORT=443
PROMETHEUS_METRICS_ADAPTER_PORT_443_TCP=tcp://10.222.119.79:443
GRAFANA_DEX_AUTHENTICATOR_PORT_443_TCP_PROTO=tcp
SHLVL=1
PROMETHEUS_METRICS_ADAPTER_SERVICE_HOST=10.222.119.79
CONTROL_PLANE_PROXY_PORT_8080_TCP_PORT=8080
TRICKSTER_PORT=tcp://10.222.85.133:443
TRICKSTER_SERVICE_PORT=443
HOME=/
CONTROL_PLANE_PROXY_SERVICE_PORT_HTTPS_METRICS=8080
CONTROL_PLANE_PROXY_PORT_8080_TCP_PROTO=tcp
CONTROL_PLANE_PROXY_SERVICE_PORT=8080
CONTROL_PLANE_PROXY_PORT=tcp://10.222.191.230:8080
GRAFANA_DEX_AUTHENTICATOR_SERVICE_PORT_HTTPS=443
PROMETHEUS_METRICS_ADAPTER_SERVICE_PORT=443
GRAFANA_DEX_AUTHENTICATOR_PORT_443_TCP=tcp://10.222.133.254:443
PROMETHEUS_METRICS_ADAPTER_PORT=tcp://10.222.119.79:443
CONTROL_PLANE_PROXY_PORT_8080_TCP=tcp://10.222.191.230:8080
GRAFANA_DEX_AUTHENTICATOR_SERVICE_HOST=10.222.133.254
TERM=xterm
POD_NAME=prometheus-main-0
KUBERNETES_PORT_443_TCP_ADDR=10.222.0.1
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
SHARD=0
GRAFANA_DEX_AUTHENTICATOR_PORT=tcp://10.222.133.254:443
KUBERNETES_PORT_443_TCP_PORT=443
GRAFANA_DEX_AUTHENTICATOR_SERVICE_PORT=443
KUBERNETES_PORT_443_TCP_PROTO=tcp
TRICKSTER_PORT_443_TCP_ADDR=10.222.85.133
TRICKSTER_PORT_443_TCP_PORT=443
TRICKSTER_PORT_443_TCP_PROTO=tcp
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT_443_TCP=tcp://10.222.0.1:443
PROMETHEUS_METRICS_ADAPTER_PORT_443_TCP_ADDR=10.222.119.79
KUBERNETES_SERVICE_HOST=10.222.0.1
PWD=/
PROMETHEUS_METRICS_ADAPTER_PORT_443_TCP_PORT=443
PROMETHEUS_METRICS_ADAPTER_PORT_443_TCP_PROTO=tcp
TRICKSTER_SERVICE_PORT_HTTPS=443
TRICKSTER_PORT_443_TCP=tcp://10.222.85.133:443
TRICKSTER_SERVICE_HOST=10.222.85.133
```

The list of valid ones are: shard, hostname, pod_name. We probably need to stick to these three variables only.

Reply:
It's rather about the envs from prometheus container, variable expanding happens in runtime as I can [see](https://github.com/prometheus/prometheus/blob/86a7064dcfc0828a10520888a188c46731986b3c/config/config.go#L80-L101).
And I'm not sure I understand why user can't use any variable, even non-existent, in such case the value would be just an empty string.

Reply:
Answered it there
> https://github.com/deckhouse/deckhouse/pull/4968#discussion_r1234966263

They can use it, of course, but it doesn't make sense. We should not encourage users to use variables we don't want them to use.

Reply:
I couldn't come up with a decent pattern for openapi schema because re2 library doesn't support negative lookahead/lookbehind syntax, so I had to remove the pattern field